pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,83 @@
+// Package multierrgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	WithContext(ctx context.Context) context.Context
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel context.CancelFunc
+	logger *zap.Logger
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger) Group {
+	return &group{
+		cancel: nil, // Set separately by WithContext
+		logger: logger,
+		wg:     sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+}
+
+// WithContext returns a new Group with a associated Context.
+// The context will be canceled if any goroutine returns an error.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext
+func (g *group) WithContext(ctx context.Context) context.Context {
+	ctx, g.cancel = context.WithCancel(ctx)
+	return ctx
+}
+
+// Wait blocks until all goroutines have completed.
+//
+// All errors returned from the goroutines will be combined into one using multierr and returned from this method.
+func (g *group) Wait() error {
+	g.wg.Wait()
+	if g.cancel != nil {
+		g.cancel()
+	}
+	return g.err
+}
+
+// Go calls the function in a new goroutine.
+// If a non-nil errors is returned, the context is canceled and
+// the error is collected using multierr and will be returned by Wait.
+func (g *group) Go(name string, f func(logger *zap.Logger) error) {
```

Comment:
Might be good to catch panics here as well. WDYT?

Reply:
Implemented in 2c9ee71.